### DISCUSSION TITLE
what is the buffering of bufferInsert() in write node during upsert?

### DISCUSSION THREAD

**Discussion by ron-gilaad** [milvus-io/milvus#37375] [Q&A and General discussion]
Hi, in the logs of Milvus when executing upsert, I see the printings of:

[writebuffer/write_buffer.go:539] ["**add growing segment**"] 
[segments/segment.go:311] ["**create segment**"]
[writebuffer/l0_write_buffer.go:229] ["**Add a new level zero segment**"]  

These printings I assume originated by **BufferData()** which was called by **writeNode Operate()**
The **BufferData()** interface is implemented by two different buffer types: **bfWriteBuffer** and **l0WriteBuffer**.  
Both implementations calls **bufferInsert()** to: buffer insert data and add segment if not exists. 

Can you please explain **what is this buffering that bufferInsert does?** and what is **segBuf** in it?
Is it the growing segment? and if so: why the data is inserted again in InsertNode later in the code 
in the **SegmentGrowingImpl.Insert()** CPP code. 

Is this the case:
**InsertBuffer** handles initial buffering.
**InsertNode/SegmentGrowingImp**l integrates data into the segment in an organized format.
**Delegator** ensures the final steps to make the segment fully available for querying.

**Comment by xiaofan-luan** [comment]
There are acaully two parts of the growing data. One on Querynode, which is SegmentGrowingImpl. The other part is on datanode, which is write buffer you've seen.

Ideally there should be only one place hold all the growing data, that's why we introduce streaming node which handles both growing segment read and flush at one place.  But right now the new written is duplicated in both querynode and datanode

**Reply by ron-gilaad** [reply]
@xiaofan-luan thanks for the answer,  I have some more questions to make it clear:

1) So the log : **[writebuffer/write_buffer.go:539] ["add growing segment"]**  - does **not** create the final growing segment which will be used for query, and this is just a temporary buffer? Also  **segBuf** in this code is actually not the final growing segment? If so, I assume that the **delegator** in **processInsert()** is the one that create the **final growing segment**. Is it true?

2) I think that **write buffer** which includes (**BufferData()** and **BufferInsert()**) - is part of **write node** - (Its **Operate()** func calls BufferData())? Is it so?

3) Is it true that **ProcessInsert()** in the **delegator** initiates and manages the insertion process, creating or managing the segments as needed, while **SegmentGrowingImpl.Insert()** performs the actual data insertion on a low level in C++ within the growing segment. The delegator essentially delegates the final insertion to the backend growing segment, where SegmentGrowingImpl.Insert() is executed to place the data in memory?

4) Where in the code is the registration of the nodes that participate in the flow graph for upsert? Is this the correct flow of upsert: **Proxy** → **DataCoord** → **Data Node** → **Write Node** (implicit, responsible for **WriteBuffer** calls) → **Insert/Delete Nodes** → **QueryNode**?

5) Delete shall be executed before Insert in upsert. But according to the log it looks as if Insert is executed before Delete, is it true? 

Thank you very much!
